sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala

```diff
@@ -123,8 +123,16 @@ case class DataSource(
     }
   }
 
-  /** Returns a source that can be used to continually read data. */
-  def createSource(): Source = {
+  /**
+   * Returns a source that can be used to continually read data.
+   *
+   * Before running a real query (e.g., df.explain), `sourceId` and `checkpointLocation` is None
+   * as they are unknown. [[ContinuousQueryManager]] should set `sourceId` and `checkpointLocation`
+   * before starting a query.
+   */
+  def createSource(
+      sourceId: Option[Long] = None,
+      checkpointLocation: Option[String] = None): Source = {
```

Comment:
Yeah, and we also don't really need to create a source there (we only need to know the schema).  Perhaps getting the schema should be separated from getting the source (like we do in FileFormat).